## Задание 4.3

Напишите функцию brackets(line), которая определяет, является ли последовательность из круглых скобок line правильной.

**Примечание 1.** Какая последовательность скобок правильная?

Правильной скобочной последовательностью назовём такую последовательность скобок, в которой для каждой открывающей скобки есть последующая соответствующая ей закрывающая скобка. Соответственно, остальные скобочные последовательности назовём неправильными. Пустую строку будем считать правильной последовательностью.

**Примечание 2.** Для решения этой задачи потребуется использовать стек.

Посимвольно переберите строку. Если встретилась открывающаяся скобка, положите её в стек. Если встретилась закрывающаяся скобка, извлеките скобку из стека.

- Если стек пустой, то есть извлечь скобку нельзя, последовательность неправильная.
- Если строка закончилась и стек стал пустым, последовательность правильная.
- Если в стеке остались скобки, последовательность неправильная.


In [15]:
# импортируем deque, чтобы одновременно использовать функционал очереди и стека
from collections import deque


# решаем задачу при помощи функции
def brackets(line):
    # с очередью можно реализовать более легкий подход для решения задачи
    # создаю очередь, чтобы засовывать и убирать скобки
    dq = deque()
    # если введенная строка пустая, то возвращаем True
    if line == '':
        return True       
    # рассматриваем каждый символ
    for symbol in line:
        if symbol == '(':
            dq.append('(')
        if symbol == ')':
            # возвращаем False если для закрывающей скобки
            # нет соответствующей 
            # открывающей скобки
            if len(dq) == 0:
                return False
            # иначе все в порядке
            else:
                dq.pop()
    # возвращаем True, еси количество открывающих скобок
    # равна количеству закрывающих
    if len(dq) == 0:
        return True
    # проверка на количество элементов в очереди даст нам ответ на вопрос
    # а есть ли для открывающей скобки соответствуюая закрывающая скобка          
    if len(dq) >= 1:
        return False    
        

print(brackets("(()())"))
# True
print(brackets(""))
# True
print(brackets("(()()))"))
# False



True
True
False


In [16]:
# запоминае как работает .pop() и .append() 
dq = deque()
dq.append('(')
dq.append('(')
print(dq)
dq.pop()
dq.pop()
print(dq)

deque(['(', '('])
deque([])


## Задание 4.4

В импортируемых в тренировочном редакторе кода переменных center, south и north хранятся списки из перечней купленных позиций в трёх торговых точках, расположенных в разных районах города.

Вначале избавьтесь от излишней вложенности: в каждой переменной (center, south, north) должен храниться объединённый список купленных товаров без разбиения по чекам.

Пример: [['Milk', 'Bread'], ['Meat']] -> ['Milk', 'Bread', 'Meat']

После этого определите, **в каком магазине было куплено больше всего товаров.**

Задание выполнено в тренировочном редакторе кода 

In [ ]:
# Это тренировочный редактор кода. Вы можете ввести любой код ниже.
# После этого нажмите кнопку «Выполнить», чтобы увидеть результат.

# Следующая строка необходима для системной работы редактора.
# В ней импортируются объекты, которые могут вам понадобиться при выполнении некоторых заданий в процессе прохождения курса.
# Не удаляйте этот импорт!
from hidden import *
from collections import Counter
from collections import deque
from collections import OrderedDict


dq = deque()
for check in center:
    dq.append(check)
dq_v2 = deque()
for elem in dq:
    dq_v2.extend(elem)
print(len(dq_v2))

dq_south = deque()
for check in south:
    dq_south.append(check)
dq_v2_south = deque()
for elem in dq_south:
    dq_v2_south.extend(elem)
print(len(dq_v2_south))   

# пополняем deque списками-чеками
dq_north = deque()
# получаем список со вложенными списками
for check in north:
    dq_north.append(check)
# создаем пустую очередь для выпрямления воженных списков
dq_v2_north = deque()
for elem in dq_north:
    dq_v2_north.extend(elem)
# смотрим на количество элементов для сравнения
print(len(dq_v2_north))    

# узнаем какое количество каждого товара в форме словаря 
center_count = Counter(dq_v2)
south_count = Counter(dq_v2_south)
north_count = Counter(dq_v2_north)

# создаем из словаря список с кортежами, чтобы отсортировать
# в порядке возрастания
north_tuple = north_count.items()
# преврашаем обьект dict_items в список для удобства
north_tuple = list(north_tuple)
north_tuple.sort(key=lambda x: x[1])
# смотрим на товар который покупали мегьше всего
print(north_tuple[0][1])

# смотрим сколько раз был куплен Bread
print(center_count['Bread'])
print(north_count['Bread'])

# создаю пустую очередь, чтобы пополнить ее популярным товаром
popular_dq = deque()
# создаю три цикла, чтобы узнать в каком списке есть товар который приобретали больше чем тот же самый товар вместе в двух других списках
for elem in center_count:
    if center_count[elem] > south_count[elem] + north_count[elem]:
        popular_dq.append(elem)
print(popular_dq)  

for elem in south_count:
    if south_count[elem] > center_count[elem] + north_count[elem]:
        popular_dq.append(elem)
print(popular_dq)

for elem in north_count:
    if north_count[elem] > center_count[elem] + south_count[elem]:
        popular_dq.append(elem)
print(popular_dq)        

# находим общее количество покупок товаров во всех магазинах вместе
all_count = center_count + south_count + north_count    
# чтобы достать значения во вложенных кортежах, помещаем все в переменную
all_tuple = all_count.most_common(2)
print(all_tuple[1][1])




## Задание 4.9

Дан список кортежей ratings, где каждый кортеж содержит название и рейтинг кафе.

Необходимо отсортировать список по убыванию рейтинга. Если рейтинги совпадают, то кафе должны быть дополнительно отсортированы в алфавитном порядке.

После сортировки создайте упорядоченный словарь cafes, используя модуль collections. Ключами должны быть названия кафе, а значениями — их рейтинги.

Используйте следующий список с рейтингами кафе:


In [17]:
from collections import deque
from collections import defaultdict
from collections import OrderedDict


ratings = [
('Old York', 3.3), ('New Age', 4.6), ('Old Gold', 3.3), ('General Foods', 4.8),
('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1),
('WokToWork', 4.9), ('WokAndRice', 4.9), ('Old Wine Cellar', 3.3), ('Nice Cakes', 3.9)
]

ratings.sort(key=lambda x: -x[1])
print(ratings)

dq = deque()
restraunt_dict = defaultdict(list)
for restraunt, rating in ratings:
    restraunt_dict[rating].append(restraunt)
print(restraunt_dict)    
# рассмариваем значение каждого элемента и сортируем его по алфавиту
for elem in restraunt_dict:
    restraunt_dict[elem].sort()
    
print(restraunt_dict)    

print(restraunt_dict.items())
# соединяем все в одной очереди, чтобы потом все собрать в словарь    
for elem in restraunt_dict:
    dq.extend(restraunt_dict[elem])        
print(dq)

# создаем упорядоченный словарь из исходного списка, 
# чтобы брать от туда значения рейтинга и перемозать новый словарь
ratings_dict = OrderedDict(ratings)
print(ratings_dict)
# пустой список нужен для хранения кортежей в которых первый элемент имя,
# а второй рейтинг
sorted_ratings = list()
# проходимся по сортированной очереди и берем значения из орингинального 
# словаря
for elem in dq:
    # в список будем поещать будущие кортежи для дальнейшего трансфера
    correct_list = list()
    # вытаскиваем значения
    if elem in ratings_dict:
        # помещаем имя
        correct_list.append(elem)
        # и добавляем его рейтинг
        correct_list.append(ratings_dict[elem])
    # отправляем готовый список в форме кортежа в общий список        
    sorted_ratings.append(tuple(correct_list))
    # персоздаем и подгатавливаем список для новой итерации
    correct_list = list()    
        
print(sorted_ratings)        
        
# создаю упорядоченный словарь
cafes = OrderedDict(sorted_ratings)
print(cafes)        
        

[('WokToWork', 4.9), ('WokAndRice', 4.9), ('General Foods', 4.8), ('New Age', 4.6), ('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1), ('Nice Cakes', 3.9), ('Old York', 3.3), ('Old Gold', 3.3), ('Old Wine Cellar', 3.3)]
defaultdict(<class 'list'>, {4.9: ['WokToWork', 'WokAndRice'], 4.8: ['General Foods'], 4.6: ['New Age'], 4.5: ['Belissimo'], 4.2: ['CakeAndCoffee', 'CakeOClock'], 4.1: ['CakeTime'], 3.9: ['Nice Cakes'], 3.3: ['Old York', 'Old Gold', 'Old Wine Cellar']})
defaultdict(<class 'list'>, {4.9: ['WokAndRice', 'WokToWork'], 4.8: ['General Foods'], 4.6: ['New Age'], 4.5: ['Belissimo'], 4.2: ['CakeAndCoffee', 'CakeOClock'], 4.1: ['CakeTime'], 3.9: ['Nice Cakes'], 3.3: ['Old Gold', 'Old Wine Cellar', 'Old York']})
dict_items([(4.9, ['WokAndRice', 'WokToWork']), (4.8, ['General Foods']), (4.6, ['New Age']), (4.5, ['Belissimo']), (4.2, ['CakeAndCoffee', 'CakeOClock']), (4.1, ['CakeTime']), (3.9, ['Nice Cakes']), (3.3, ['Old Gold', 'Old Wine Cellar', '

In [ ]:
from collections import deque
from collections import defaultdict
from collections import OrderedDict


ratings = [
('Old York', 3.3), ('New Age', 4.6), ('Old Gold', 3.3), ('General Foods', 4.8),
('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1),
('WokToWork', 4.9), ('WokAndRice', 4.9), ('Old Wine Cellar', 3.3), ('Nice Cakes', 3.9)
]

ratings.sort(key=lambda x: -x[1])
# в пустой deque будем собирать правельную очередность имен ресторанов
dq = deque()
# создаем словарь у которого список со значениями
# для дальнейшей сортировки по алфавиту
restraunt_dict = defaultdict(list)
for restraunt, rating in ratings:
    restraunt_dict[rating].append(restraunt)    
# рассмариваем значение каждого элемента и сортируем его по алфавиту
for elem in restraunt_dict:
    restraunt_dict[elem].sort()
# соединяем все в одной очереди, чтобы потом все собрать в словарь    
for elem in restraunt_dict:
    dq.extend(restraunt_dict[elem])        
# создаем упорядоченный словарь из исходного списка, 
# чтобы брать от туда значения рейтинга и пересоздать новый словарь
ratings_dict = OrderedDict(ratings)
# пустой список нужен для хранения кортежей в которых первый элемент имя,
# а второй рейтинг
sorted_ratings = list()
# проходимся по сортированной очереди и берем значения из орингинального 
# словаря
for elem in dq:
    # в список будем поещать будущие кортежи для дальнейшего трансфера
    correct_list = list()
    # вытаскиваем значения
    if elem in ratings_dict:
        # помещаем имя
        correct_list.append(elem)
        # и добавляем его рейтинг
        correct_list.append(ratings_dict[elem])
    # отправляем готовый список в форме кортежа в общий список        
    sorted_ratings.append(tuple(correct_list))
    # персоздаем и подгатавливаем список для новой итерации
    correct_list = list()    
        
        
# создаю упорядоченный словарь
cafes = OrderedDict(sorted_ratings)
#print(cafes) 

OrderedDict({'WokAndRice': 4.9, 'WokToWork': 4.9, 'General Foods': 4.8, 'New Age': 4.6, 'Belissimo': 4.5, 'CakeAndCoffee': 4.2, 'CakeOClock': 4.2, 'CakeTime': 4.1, 'Nice Cakes': 3.9, 'Old Gold': 3.3, 'Old Wine Cellar': 3.3, 'Old York': 3.3})


## Задание 4.10

Напишите функцию task_manager(tasks), которая принимает список задач tasks для нескольких серверов. Каждый элемент списка состоит из кортежа (<номер задачи>, <название сервера>, <высокий приоритет задачи>).

Функция должна создавать словарь и заполнять его задачами по следующему принципу: название сервера — ключ, по которому хранится очередь задач для конкретного сервера.

Если поступает задача без высокого приоритета (последний элемент кортежа — False), добавить номер задачи в конец очереди. Если приоритет высокий, добавить номер в начало.

Для словаря используйте defaultdict, для очереди — deque.

Функция возвращает полученный словарь с задачами.

In [ ]:
# получаю функционал из collections
from collections import defaultdict
from collections import deque


tasks = [(36871, 'office', False),
(40690, 'office', False),
(35364, 'voltage', False),
(41667, 'voltage', True),
(33850, 'office', False)]


# пишем текс функции
def task_manager(tasks):
    # создаем обьект defaultdict для последующего заполнения
    # обьект deque может быть пополняем спереди
    tasks_dict = defaultdict(deque)
    # смотрим на каждый элемент
    for number, name, priority in tasks:
        # определяем с какой стороны добавлять
        if priority == False:
            # создаем постоянно обновляемую деку по назначению
            exampe_deque = deque()
            # пополняем
            exampe_deque.append(number)
            # расширяем словарь по необходимому ключу
            tasks_dict[name].extend(exampe_deque)
        # определяем с какой стороны добавлять
        if priority == True:
            exampe_deque = deque()
            exampe_deque.append(number)
            tasks_dict[name].extendleft(exampe_deque)

    return tasks_dict

print(task_manager(tasks))    
    
    

defaultdict(<class 'collections.deque'>, {'office': deque([36871, 40690, 33850]), 'voltage': deque([41667, 35364])})
